File: src/main/java/org/algohub/engine/judge/Function.java
package org.algohub.engine.judge;

import com.fasterxml.jackson.annotation.JsonCreator;
import com.fasterxml.jackson.annotation.JsonIgnoreProperties;
import com.fasterxml.jackson.annotation.JsonProperty;

@JsonIgnoreProperties(ignoreUnknown = true)
public class Function {

    private final String name;
    /**
     * Return metadata.
     */
    @JsonProperty("return")
    private final Return returnStatement;
    /**
     * Parameters' metadata.
     */
    private final Parameter[] parameters;

    /**
     * Since this class is immutable, need to provide a method for Jackson.
     */
    @JsonCreator
    public Function(@JsonProperty("name") final String name,
                    @JsonProperty("return") final Return returnStatement,
                    @JsonProperty("parameters") final Parameter[] parameters) {
        this.name = name;
        this.returnStatement = returnStatement;
        this.parameters = parameters;
    }

    public String getName() {
        return name;
    }

    public Return getReturnStatement() {
        return returnStatement;
    }

    public Parameter[] getParameters() {
        return parameters;
    }

    /**
     * Return type.
     */
    @JsonIgnoreProperties(ignoreUnknown = true)
    public static class Return {
        /**
         * Return data type.
         */
        private final String type;
        /**
         * Comment of returned value.
         */
        private final String comment;

        /**
         * Since this class is immutable, need to provide a method for Jackson.
         */
        @JsonCreator
        public Return(@JsonProperty("type") final String type,
                      @JsonProperty("comment") final String comment) {
            this.type = type;
            this.comment = comment;
        }

        public String getType() {
            return type;
        }

        public String getComment() {
            return comment;
        }
    }

    /**
     * Function parameters' metadata.
     */
    @JsonIgnoreProperties(ignoreUnknown = true)
    public static class Parameter {
        /**
         * Parameter name.
         */
        private final String name;
        /**
         * Parameter type.
         */
        private final String type;
        /**
         * Parameter comment.
         */
        private final String comment;

        /**
         * Since this class is immutable, need to provide a method for Jackson.
         */
        @JsonCreator
        public Parameter(@JsonProperty("name") final String name,
                         @JsonProperty("type") final String type, @JsonProperty("comment") final String comment) {
            this.name = name;
            this.type = type;
            this.comment = comment;
        }

        public String getName() {
            return name;
        }

        public String getType() {
            return type;
        }

        public String getComment() {
            return comment;
        }
    }
}

File: src/main/kotlin/org/algohub/engine/judge/Function.kt
package org.algohub.engine.judge

import com.fasterxml.jackson.annotation.JsonCreator
import com.fasterxml.jackson.annotation.JsonIgnoreProperties
import com.fasterxml.jackson.annotation.JsonProperty

@JsonIgnoreProperties(ignoreUnknown = true)
class Function
/**
 * Since this class is immutable, need to provide a method for Jackson.
 */
@JsonCreator
constructor(@JsonProperty("name") val name: String,
            /**
             * Return metadata.
             */
            @JsonProperty("return")
            val returnStatement: Function.Return,
            /**
             * Parameters' metadata.
             */
            @JsonProperty("parameters") val parameters: Array<Function.Parameter>) {

    /**
     * Return type.
     */
    @JsonIgnoreProperties(ignoreUnknown = true)
    class Return
    /**
     * Since this class is immutable, need to provide a method for Jackson.
     */
    @JsonCreator
    constructor(
            /**
             * Return data type.
             */
            @JsonProperty("type") val type: String,
            /**
             * Comment of returned value.
             */
            @JsonProperty("comment") val comment: String)

    /**
     * Function parameters' metadata.
     */
    @JsonIgnoreProperties(ignoreUnknown = true)
    class Parameter
    /**
     * Since this class is immutable, need to provide a method for Jackson.
     */
    @JsonCreator
    constructor(
            /**
             * Parameter name.
             */
            @JsonProperty("name") val name: String,
            /**
             * Parameter type.
             */
            @JsonProperty("type") val type: String,
            /**
             * Parameter comment.
             */
            @JsonProperty("comment") val comment: String)
}
